File: 백준/Platinum/9202. Boggle/Boggle.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

// 9202
public class Main {
    static TrieNode root = new TrieNode();
    static int w, b;
    static char[][] map;
    static boolean[][] visited;
    static String answer;
    static int sum;
    static int count;
    static StringBuilder sb;

    static int[] dx = {1, -1, 0, 0, 1, 1, -1, -1};
    static int[] dy = {0, 0, 1, -1, 1, -1, 1, -1};
    static int[] score = {0, 0, 0, 1, 1, 2, 3, 5, 11};

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        w = Integer.parseInt(br.readLine());
        for (int i = 0; i < w; i++) {
            insertTrieNode(br.readLine());
        }
        br.readLine();
        b = Integer.parseInt(br.readLine());
        StringBuilder resultSb = new StringBuilder();
        for (int i = 0; i < b; i++) {
            map = new char[4][4];
            visited = new boolean[4][4];
            answer = "";
            sum = 0;
            count = 0;
            sb = new StringBuilder();

            for (int j = 0; j < 4; j++) {
                String in = br.readLine();
                for (int k = 0; k < 4; k++) {
                    map[j][k] = in.charAt(k);
                }
            }
            br.readLine();
            for (int x = 0; x < 4; x++) {
                for (int y = 0; y < 4; y++) {
                    // 출발 가능 조건 -> root 가 해당 child 를 가지면
                    if (root.hasChild(map[x][y])) {
                        search(x, y, root.getChild(map[x][y]));
                    }
                }
            }
            // 결과 출력
            resultSb.append(sum + " " + answer + " " + count + "\n");
            root.clearHit();
        }
        System.out.println(resultSb.toString());
    }

    static void search(int x, int y, TrieNode node) {
        // 1. 체크인 -> visited
        visited[x][y] = true;
        sb.append(map[x][y]);
        // 2. 목적지에 도달하였는가 -> isWord == true, isHit == false
        if (node.isWord && !node.isHit) {
            node.isHit = true;
            // 답 -> 길이, 단어,
            String foundWord = sb.toString();
            int length = foundWord.length();
            count++;
            sum += score[length];
            if (answer.length() < length) {
                answer = foundWord;
            } else if (answer.length() == length) {
                if (answer.compareTo(foundWord) > 0) {
                    answer = foundWord;
                }
            }
        }
        // 3. 연결된 곳을 순회 -> 8방
        for (int i = 0; i < 8; i++) {
            int tx = x + dx[i];
            int ty = y + dy[i];
            // 4. 가능한가? -> map 경계, 방문하지 않았는지, node 가 해당 자식을 가지고 있는지
            if (0 <= tx && tx < 4 && 0 <= ty && ty < 4) {
                if (visited[tx][ty] == false && node.hasChild(map[tx][ty])) {
                    // 5. 간다
                    search(tx, ty, node.getChild(map[tx][ty]));
                }
            }
        }

        // 6. 체크아웃
        sb.deleteCharAt(sb.length() - 1);
        visited[x][y] = false;
    }

    static void insertTrieNode(String word) {
        TrieNode current = root;
        for (int i = 0; i < word.length(); i++) {
            char a = word.charAt(i);
            if (current.child[a - 'A'] == null) {
                current.child[a - 'A'] = new TrieNode();
            } // 자식이 생김
            current = current.child[a - 'A'];
        }
        // 단어 완성
        current.isWord = true;
    }
}

class TrieNode {
    boolean isWord = false;
    boolean isHit = false;
    TrieNode[] child = new TrieNode[26];

    void clearHit() {
        isHit = false;
        for (int i = 0; i < child.length; i++) {
            if (child[i] != null) {
                child[i].clearHit();
            }
        }
    }

    boolean hasChild(char c) {
        return child[c - 'A'] != null;
    }

    TrieNode getChild(char c) {
        return child[c - 'A'];
    }
}

File: 백준/Platinum/9202. Boggle/Boggle.py
import sys

input = lambda: sys.stdin.readline().rstrip()

dx = [-1,1,0,0,-1,1,-1,1]
dy = [0,0,-1,1,-1,-1,1,1]



w = int(input())
word_dictionary = []
for _ in range(w):
    word_dictionary.append(input().strip())
input()




def solution():
  b = int(input())

  trie = make_trie()

  for _ in range(b):
    board = [list(input()) for _ in range(4)]
    input()
    search_result = set()
    visited = [[False] * 4 for _ in range(4)]
    for x in range(4):
      for y in range(4):
        now = trie
        for child in now.child:
          if child and board[x][y] == child.alpha:
            visited[x][y] = True
            search_word(child, x, y, board, board[x][y], visited, search_result)
            visited[x][y] = False
            break
    long_word = ''
    score = 0
    search_result = list(search_result)
    search_result.sort()
    for word in search_result:
      if len(long_word) < len(word):
        long_word = word
      if 3 <= len(word) <= 4:
        score += 1
      elif len(word) == 5:
        score += 2
      elif len(word) == 6:
        score += 3
      elif len(word) == 7:
        score += 5
      elif len(word) == 8:
        score += 11
    print(score, long_word, len(search_result))
    
def search_word(cur, x, y, board, word, visited, search_result):
  if cur.isWord:
    search_result.add(word)

  if len(word) == 8:
    return
  
  for i in range(8):
    nx = x + dx[i]
    ny = y + dy[i]

    if nx >= 0 and nx < 4 and ny >= 0 and ny < 4 and not visited[nx][ny]:
      for child in cur.child:
        if child and child.alpha == board[nx][ny]:
          visited[nx][ny] = True
          search_word(child, nx, ny, board, word + board[nx][ny], visited, search_result)
          visited[nx][ny] = False
  

def make_trie():
  trie = Node()

  for word in word_dictionary:
    insert_trie_node(trie, word)
  return trie

def insert_trie_node(root, word):
  cur = root
  for wo in word:
    flag = False
    for child in cur.child:
      if wo == child.alpha:
        flag = True
        cur = child
        break
    if not flag:
      node = Node(wo)
      cur.child.append(node)
      cur = node
  cur.isWord = True

class Node:
  def __init__(self, alpha = ''):
    self.alpha = alpha
    self.child = []
    self.isWord = False
  
solution()